# Data Quality Report - Before Cleaning

This section evaluates the dataset before cleaning to identify:
- Missing values
- Duplicate records
- Incorrect data types
- Invalid values
- Potential outliers

In [7]:
df.shape


(10000, 8)

In [8]:
df.columns

Index(['Transaction ID', 'Item', 'Quantity', 'Price Per Unit', 'Total Spent',
       'Payment Method', 'Location', 'Transaction Date'],
      dtype='str')

In [9]:
df.isnull().sum()

Transaction ID         0
Item                 333
Quantity             138
Price Per Unit       179
Total Spent          173
Payment Method      2579
Location            3265
Transaction Date     159
dtype: int64

## Missing Values

The dataset was checked for missing values. Columns with missing values require a cleaning strategy based on the business meaning of the column.

In [10]:
df.duplicated().sum()

np.int64(0)

## Duplicate Records

The dataset contains X duplicate rows. These records will be removed because duplicate transactions can distort analysis.

In [11]:
df.dtypes

Transaction ID      str
Item                str
Quantity            str
Price Per Unit      str
Total Spent         str
Payment Method      str
Location            str
Transaction Date    str
dtype: object

In [14]:
df["Transaction Date"].isnull().sum()

np.int64(159)

In [15]:
df = df.dropna(subset=["Transaction Date"])

### Transaction Date

159 records were missing transaction dates. These rows were removed because transaction dates are required for time-based analysis, and the missing values could not be reliably inferred.

In [16]:
df["Quantity"].isnull().sum()

np.int64(138)

In [6]:
import pandas as pd

# Use the dataframe already defined in this notebook.
# If it is not defined, try to load a CSV from the project folder.
if "df" not in globals():
	try:
		df = pd.read_csv("data.csv")
	except FileNotFoundError:
		try:
			df = pd.read_csv("transactions.csv")
		except FileNotFoundError:
			# Fallback so the notebook does not crash if the data file is not available yet.
			df = pd.DataFrame({"Quantity": []})

if not isinstance(df, pd.DataFrame):
	raise ValueError("The variable 'df' is not a pandas DataFrame. Load the dataset before converting 'Quantity'.")

df["Quantity"] = pd.to_numeric(df["Quantity"], errors="coerce")
quantity_median = df["Quantity"].median()

print(quantity_median)

nan


In [7]:
quantity_median = df["Quantity"].median()

quantity_median

nan

In [8]:
df["Quantity"].unique()

array([], dtype=float64)

In [9]:
print(df.shape)

(0, 1)


In [10]:
print(df.columns)

Index(['Quantity'], dtype='str')


In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 0 entries
Data columns (total 1 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Quantity  0 non-null      float64
dtypes: float64(1)
memory usage: 132.0 bytes


In [12]:
df.head()

,Quantity


In [18]:
df["Quantity"].dtype

<StringDtype(storage='python', na_value=nan)>

In [19]:
df["Quantity"] = pd.to_numeric(
    df["Quantity"],
    errors="coerce"
)